packages/grafana-ui/src/components/Table/TableNG/utils.ts

```diff
@@ -69,19 +48,89 @@ export function getDefaultRowHeight(theme: GrafanaTheme2, cellHeight?: TableCell
   return TABLE.CELL_PADDING * 2 + bodyFontSize * lineHeight;
 }
 
+/**
+ * @internal
+ * Returns true if cell inspection (hover to see full content) is enabled for the field.
+ */
+export function isCellInspectEnabled(field: Field): boolean {
+  return field.config?.custom?.inspect ?? false;
+}
+
+/**
+ * @internal
+ * Returns true if text wrapping should be applied to the cell.
+ */
+export function shouldTextWrap(field: Field): boolean {
+  const cellOptions = getCellOptions(field);
+  return ('wrapText' in cellOptions && cellOptions.wrapText) ?? false;
+}
+
+// matches characters which CSS
+const spaceRegex = /[\s-]/;
+
+export interface GetMaxWrapCellOptions {
+  colWidths: number[];
+  avgCharWidth: number;
+  wrappedColIdxs: boolean[];
+}
+
+/**
+ * @internal
+ * loop through the fields and their values, determine which cell is going to determine the
+ * height of the row based on its content and width, and then return the text, index, and number of lines for that cell.
+ */
+export function getMaxWrapCell(
+  fields: Field[],
+  rowIdx: number,
+  { colWidths, avgCharWidth, wrappedColIdxs }: GetMaxWrapCellOptions
+): {
+  text: string;
+  idx: number;
+  numLines: number;
+} {
+  let maxLines = 1;
+  let maxLinesIdx = -1;
+  let maxLinesText = '';
+
+  for (let i = 0; i < colWidths.length; i++) {
+    if (wrappedColIdxs[i]) {
+      const field = fields[i];
+      // special case: for the header, provide `-1` as the row index.
+      const cellTextRaw = rowIdx === -1 ? getDisplayName(field) : field.values[rowIdx];
+
+      if (cellTextRaw != null) {
+        const cellText = String(cellTextRaw);
+
+        if (spaceRegex.test(cellText)) {
+          const charsPerLine = colWidths[i] / avgCharWidth;
+          const approxLines = cellText.length / charsPerLine;
+
+          if (approxLines > maxLines) {
+            maxLines = approxLines;
+            maxLinesIdx = i;
+            maxLinesText = cellText;
+          }
+        }
+      }
+    }
+  }
+
+  return { text: maxLinesText, idx: maxLinesIdx, numLines: maxLines };
+}
+
 /**
  * @internal
  * Returns true if text overflow handling should be applied to the cell.
  */
-export function shouldTextOverflow(
-  fieldType: FieldType,
-  cellType: TableCellDisplayMode,
-  textWrap: boolean,
-  cellInspect: boolean
-): boolean {
-  // Tech debt: Technically image cells are of type string, which is misleading (kinda?)
-  // so we need to ensure we don't apply overflow hover states fo type image
-  return fieldType === FieldType.string && cellType !== TableCellDisplayMode.Image && !textWrap && !cellInspect;
+export function shouldTextOverflow(field: Field): boolean {
```

Comment:
this is a preference thing for utils generally, but since methods like `shouldTextWrap` or `isCellInspectEnabled` do inexpensive lookups on the field, I prefer utils which pass the whole field in rather than the outputs of those checks so that the method is more composable in any context without additional legwork. I find this to be more reusable and less prone to mistakes in the long run.

Reply:
another way to do it if the checks do become expensive might be

```ts
export function shouldTextOverflow(
  field: Field,
  textWrap = shouldTextWrap(field),
  // etc...
) {
  // ...
}
```

so that the user can provide a pre-calculated value optionally to avoid re-calculating, but it's not strictly necessary and the call can just be `shouldTextOverflow(field)` in a normal use case.